bloom.c

```diff
@@ -0,0 +1,265 @@
+{
+	int i;
+	uint64_t mod = filter->len * BITS_PER_BLOCK;
+
+	if (!mod)
+		return 1;
+
+	for (i = 0; i < settings->num_hashes; i++) {
+		uint64_t hash_mod = key->hashes[i] % mod;
+		uint64_t block_pos = hash_mod / BITS_PER_BLOCK;
+		if (!(filter->data[block_pos] & get_bitmask(hash_mod)))
+			return 0;
+	}
+
+	return 1;
+}
+
+void fill_bloom_key(const char *data,
+		    int len,
+		    struct bloom_key *key,
+		    struct bloom_filter_settings *settings)
+{
+	int i;
+	uint32_t seed0 = 0x293ae76f;
+	uint32_t seed1 = 0x7e646e2c;
+
+	uint32_t hash0 = seed_murmur3(seed0, data, len);
+	uint32_t hash1 = seed_murmur3(seed1, data, len);
+
+	key->hashes = (uint32_t *)xcalloc(settings->num_hashes, sizeof(uint32_t));
+	for (i = 0; i < settings->num_hashes; i++) {
+		key->hashes[i] = hash0 + i * hash1;
+	}
+}
+
+static void add_key_to_filter(struct bloom_key *key,
+			      struct bloom_filter *filter,
+			      struct bloom_filter_settings *settings)
+{
+	int i;
+	uint64_t mod = filter->len * BITS_PER_BLOCK;
+
+	for (i = 0; i < settings->num_hashes; i++) {
+		uint64_t hash_mod = key->hashes[i] % mod;
+		uint64_t block_pos = hash_mod / BITS_PER_BLOCK;
+
+		filter->data[block_pos] |= get_bitmask(hash_mod);
+	}
+}
+
+static void fill_filter_from_graph(struct commit_graph *g,
+				   struct bloom_filter *filter,
+				   struct commit *c)
+{
+	uint32_t prev_index = 0;
+	uint32_t next_index = get_be32(g->chunk_bloom_indexes + 4 * c->graph_pos);
+
+	if (c->graph_pos > 0) {
+		prev_index = get_be32(g->chunk_bloom_indexes + 4 * (c->graph_pos - 1));
+	}
+
+	filter->len = next_index - prev_index;
+	filter->data = (uint64_t *)(g->chunk_bloom_data + 8 * prev_index + 12);
+}
+
+void load_bloom_filters(void)
+{
+	init_bloom_filter_slab(&bloom_filters);
+}
+
+struct bloom_filter *get_bloom_filter(struct repository *r,
+				      struct commit *c,
+				      int compute_if_null)
+{
+	struct bloom_filter *filter;
+	struct bloom_filter_settings settings = DEFAULT_BLOOM_FILTER_SETTINGS;
+	int already_computed = 0;
+
+	filter = bloom_filter_slab_at(&bloom_filters, c);
+
+	if (!filter->data) {
+		load_commit_graph_info(r, c);
+		if (c->graph_pos != COMMIT_NOT_FROM_GRAPH) {
+			if (r->objects->commit_graph->chunk_bloom_indexes) {
+				fill_filter_from_graph(r->objects->commit_graph, filter, c);
+				already_computed = 1;
+			}
+		}
+	}
+
+	if (!filter->data && compute_if_null) {
+
+		if (!already_computed) {
+			int i;
+			struct rev_info revs;
+			const char *revs_argv[] = {NULL, "HEAD", NULL};
+			init_revisions(&revs, NULL);
+			revs.diffopt.flags.recursive = 1;
+
+			setup_revisions(2, revs_argv, &revs, NULL);
+
+			if (c->parents)
+				diff_tree_oid(&c->parents->item->object.oid, &c->object.oid, "", &revs.diffopt);
+			else
+				diff_tree_oid(NULL, &c->object.oid, "", &revs.diffopt);
+			diffcore_std(&revs.diffopt);
+
+			//fprintf(stderr, "computing filter for %s:\n", oid_to_hex(&c->object.oid));
+
+			if (diff_queued_diff.nr <= 512) {
+				struct hashmap pathmap;
```

Comment:
From what I have studied of the git codebase, everyone uses the hashmap differently and there is no standard or recommended way to do this. I used the example in hashmap.h and what diff-tool and blame.h are doing to get what I wanted here. I am sure it will get reviewed accordingly on the mailing list. :)